### ISSUE TITLE
Metal: proper tracking of read/write/readwrite textures

### ISSUE THREAD

**Issue by fatkas** [KhronosGroup/SPIRV-Cross#349]
With current code MSL generator would produce incorrect shader code when write only texture is used. That seems to be because DecorationNonReadable is not propagated across all the instances of the texture. Global texture declaration ends up being access::write but all of the function parameters are access::read_write . I'm having hard times properly propagating the flags so any implementation suggestions are highly welcomed.

here's the source hlsl

```
struct Params
{
	float4 CascadeOffset;
};

cbuffer CB0: register(b0) { Params CB0; }

RWTexture3D<float4> Result: register(u0);
Texture3D<float4> Input: register(t1);

[numthreads(16, 16, 1)]
void CopyCS(uint3 tid: SV_DispatchThreadID)
{
	Result[tid + uint3(0, 0, CB0.CascadeOffset.y)] = Input[tid];
}
```

and that's the resulting msl

```
#include <metal_stdlib>
#include <simd/simd.h>

using namespace metal;

struct Params
{
    float4 CascadeOffset;
};

struct CB0
{
    Params CB0;
};

void _CopyCS(thread const uint3& tid, thread texture3d<float> Input, thread texture3d<float, access::read_write> Result, constant CB0& v_35)
{
    float4 storeTemp = Input.read(uint3(tid), 0);
    Result.write(storeTemp, uint3((tid + uint3(0u, 0u, uint(v_35.CB0.CascadeOffset.y)))));
}

kernel void main0(texture3d<float> Input [[texture(1)]], texture3d<float, access::write> Result [[texture(0)]], constant CB0& v_35 [[buffer(0)]], uint3 gl_GlobalInvocationID [[thread_position_in_grid]])
{
    uint3 tid = gl_GlobalInvocationID;
    uint3 param = tid;
    _CopyCS(param, Input, Result, v_35);
}
```

**Comment by HansKristian-ARM** [comment]
Hm, this sounds really tricky as IIRC, SPIR-V can add and remove qualifiers when passing images around. Not sure how strict MSL is about this, but this could get really hairy. Pinging @billhollings.

**Comment by zeux** [comment]
As a note, the reason why write-only vs read-write tracking is important to us is because apparently to use read-write you need a much later version of the OS (specifically, you need a much higher shader profile).

**Comment by zeux** [comment]
This is mostly addressed by #358 now. That change does not track use of objects inside functions, so if you have a function that accepts RWTexture and pass the object to this function the tracking won't work and the texture will probably still be read-write, but at least for most basic cases it should work fine.

**Comment by zeux** [comment]
>  if you have a function that accepts RWTexture and pass the object to this function the tracking won't work

Actually legalization in glslang includes exhaustive inlining so this shouldn't happen when using HLSL as a source. This can probably be closed now.
